service/role.go

```diff
@@ -145,109 +215,174 @@ func (r *roleService) getRoleToken(ctx context.Context, domain, role, proxyForPr
+func (r *roleService) fetchRoleToken(ctx context.Context, domain, role, proxyForPrincipal string, minExpiry, maxExpiry time.Duration) (*RoleToken, error) {
+	glg.Debugf("get role token, domain: %s, role: %s, proxyForPrincipal: %s, minExpiry: %s, maxExpiry: %s", domain, role, proxyForPrincipal, minExpiry, maxExpiry)
+
+	// get the n-token
+	tok, err := r.token()
+	if err != nil {
+		return nil, err
+	}
+
+	// prepare request object
+	u := r.getRoleTokenAthenzURL(domain, role, minExpiry, maxExpiry, proxyForPrincipal)
+	glg.Debugf("url: %v", u)
+	req, err := http.NewRequest(http.MethodGet, u, nil)
+	if err != nil {
+		glg.Debugf("fail to create request object, error: %s", err)
+		return nil, err
+	}
+	req.Header.Set(r.athenzPrincipleHeader, tok)
 
+	res, err := r.httpClient.Do(req.WithContext(ctx))
 	if err != nil {
 		return nil, err
 	}
 
-	return tok.(*RoleToken), nil
+	defer func() {
+		if res != nil {
+			io.Copy(ioutil.Discard, res.Body)
+			if res.Body != nil {
+				res.Body.Close()
+			}
+		}
+	}()
+
+	if res.StatusCode != http.StatusOK {
+		ioutil.ReadAll(res.Body)
+		return nil, ErrRoleTokenRequestFailed
+	}
+
+	var data *RoleToken
+	if err = json.NewDecoder(res.Body).Decode(&data); err != nil {
+		return nil, err
+	}
+
+	return data, nil
 }
 
 func (r *roleService) getCache(domain, role, principal string) (*RoleToken, bool) {
-	val, ok := r.domainRoleCache.Get(encode(domain, role))
+	val, ok := r.domainRoleCache.Get(encode(domain, role, principal))
 	if !ok {
 		return nil, false
 	}
 	return val.(*cacheData).token, ok
 }
 
-func encode(domain, role string) string {
-	return fmt.Sprintf("%s-%s", domain, role)
+func encode(domain, role, principal string) string {
+	s := []string{domain, role, principal}
+	if principal == "" {
+		return strings.Join(s[:2], cacheKeySeparater)
+	}
+	return strings.Join(s, cacheKeySeparater)
 }
 
-func decode(key string) (string, string) {
-	keys := strings.SplitN(key, "-", 3)
-	if len(keys) < 2 {
-		return key, ""
-	}
-	return keys[0], keys[1]
+func decode(key string) (string, string, string) {
+	keys := strings.SplitN(key, cacheKeySeparater, 3)
+	res := []string{"", "", ""}
+	copy(res, keys)
+	return res[0], res[1], res[2]
 }
 
-func getRoleTokenAthenzURL(athenzURL, domain, role string, minExpiry, maxExpiry time.Duration, proxyForPrincipal string) string {
-	u := fmt.Sprintf("https://%s/domain/%s/token?role=%s", strings.TrimPrefix(strings.TrimPrefix(athenzURL, "https://"), "http://"), domain, url.QueryEscape(role))
+func (r *roleService) getRoleTokenAthenzURL(domain, role string, minExpiry, maxExpiry time.Duration, proxyForPrincipal string) string {
+	u := fmt.Sprintf("https://%s/domain/%s/token?role=%s", strings.TrimPrefix(strings.TrimPrefix(r.athenzURL, "https://"), "http://"), domain, url.QueryEscape(role))
+
+	minExp := r.expiry
+	if minExpiry > 0 {
+		minExp = minExpiry
+	}
+	if minExp > 0 {
+		u += fmt.Sprintf("&minExpiryTime=%d", minExp/time.Second)
+	}
+
+	maxExp := r.expiry
```

Comment:
I remember that we have a conclusion to not have a default for max?

Reply:
The program will set to default 0 and at line 380 it will check if it is 0 and add the parameter

Reply:
u means `r.expiry == 0`, right?
I think even if `r.expiry != 0`, we should NOT send `max`
we send `max` if and only if user specified in the request?

Reply:
Umm... yes

Reply:
please check the latest commit

Reply:
👍🏻

Reply:
👍🏻 for latest commit
related: https://github.com/yahoojapan/athenz-client-sidecar/pull/15#discussion_r336348871

Reply:
fixed